### PR TITLE
Set router skip clean to false 

### PR DESCRIPTION
**Pull request by suneyz**
### Summary

This enables agent to handle request like "//v2/credential" correctly. This brings back the support for //path that agent version 1.20.3 or before
https://github.com/aws/amazon-ecs-agent/issues/1839

### Implementation details

set router skip clean to false
doc for skip clean https://godoc.org/github.com/gorilla/mux#Router.SkipClean

### Testing

New tests cover the changes: 

yes,

unit test, functional test

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by samuelkarp** [comment]
> New tests cover the changes:

I don't see any tests in the pull request.  Can you add a functional test covering this change?

**Comment by yunhee-l** [comment]
v3 tests are failing, is this causing different regression?
TestTaskHTTPEndpointErrorCode404 (0.02s)
TestTaskHTTPEndpointErrorCode404/Test_path:_/v3///task/
TestTaskHTTPEndpointErrorCode400 
TestTaskHTTPEndpointErrorCode400/Test_path:_/v3//stats 
TestTaskHTTPEndpointErrorCode400/Test_path:_/v3//task 
TestTaskHTTPEndpointErrorCode400/Test_path:_/v3//task/stats

**Comment by suneyz** [comment]
> > New tests cover the changes:
> 
> I don't see any tests in the pull request. Can you add a functional test covering this change?

Great call out. Added test file. There is actually unit test to test the router behavior.

**Comment by suneyz** [comment]
> > New tests cover the changes:
> 
> I don't see any tests in the pull request. Can you add a functional test covering this change?

Great call out! Added tests, there is actually unit tests that covers router

**Comment by samuelkarp** [comment]
Can you add a functional test as well?

**Comment by suneyz** [comment]
> Can you add a functional test as well?

updated with functional tests

File: misc/task-server-endpoint-validator/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM ubuntu:latest
```
Comment (shubham2892) [inline misc/task-server-endpoint-validator/Dockerfile]: Locking this to a version would be a good idea. We are using 16.04 at other places.

File: agent/handlers/task_server_setup_test.go [hunk 2]
```diff
@@ -1101,11 +1128,11 @@ func TestTaskHTTPEndpointErrorCode400(t *testing.T) {
 		"/v3/wrong-v3-endpoint-id",
 		"/v3/",
 		"/v3/wrong-v3-endpoint-id/stats",
-		"/v3//stats",
+		"/v3/stats",
```
Comment (sharanyad) [inline agent/handlers/task_server_setup_test.go]: Why was this changed? Shouldn't `"/v3//stats"` work as well?

File: agent/functional_tests/testdata/taskdefinitions/task-server-endpoint-validator/task-definition.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+    "family": "task-server-endpoint-validator",
+    "taskRoleArn": "$$$TASK_ROLE$$$",
+    "networkMode": "host",
+    "containerDefinitions": [{
+        "image": "127.0.0.1:51670/amazon/task-server-endpoint-validator:latest",
+        "name": "container_1",
```
Comment (shubham2892) [inline agent/functional_tests/testdata/taskdefinitions/task-server-endpoint-validator/task-definition.json]: nit: rename this to a more descripritive one.

File: agent/handlers/task_server_setup_test.go (hunk 2 above)
Comment (suneyz) [reply agent/handlers/task_server_setup_test.go]: we handle double/muti slash with a redirect prior to 1.20.3 release. at 1.21.0 release we changed the router and added the behavior to not clean double/multi slash. This is the root cause of this issue https://github.com/aws/amazon-ecs-agent/issues/1839.

File: agent/handlers/task_server_setup.go [hunk 4]
```diff
@@ -56,9 +56,9 @@ func taskServerSetup(credentialsManager credentials.Manager,
 	containerInstanceArn string) *http.Server {
 	muxRouter := mux.NewRouter()
 
-	// Set this so that for request like "/v3//metadata/task", the Agent will pass
-	// it to task metadata handler instead of returning a 301 error.
-	muxRouter.SkipClean(true)
+	// Set this to false so that for request like "//v3//metadata/task"
+	// to permanently redirect(301) to "/v3/metadata/task" handler
+	muxRouter.SkipClean(false)
```
Comment (fierlion) [inline agent/handlers/task_server_setup.go]: Since this defaults to `false` (from the docs: "SkipClean defines the path cleaning behaviour for new routes. The initial value is false.") I initially wondered if we even need the explicit setting here.  
But I like it here.  It's deliberate.  
I would make the comment clearer (like the one in the docs: "When false, the path will be cleaned, so //v3//metadata/task will become /v3/metadata/task")

Comment (suneyz) [reply agent/handlers/task_server_setup.go]: I think a 301 redirect is more clear as it is not "becoming" the new url. it is a specific 301 redirect to the new url. let me know how you think

File: misc/task-server-endpoint-validator/Dockerfile (hunk 1 above)
Comment (suneyz) [reply misc/task-server-endpoint-validator/Dockerfile]: make sense :)

File: agent/functional_tests/testdata/taskdefinitions/task-server-endpoint-validator/task-definition.json (hunk 3 above)
Comment (suneyz) [reply agent/functional_tests/testdata/taskdefinitions/task-server-endpoint-validator/task-definition.json]: do you have a suggestion? i thought the task definition and image name is descriptive enough and i didn't want to introduce similar/same name to cause confusion as it is the only container in the test.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 5]
```diff
@@ -1427,3 +1427,45 @@ func TestElasticInferenceValidator(t *testing.T) {
 	exitCode, _ := task.ContainerExitcode("container_1")
 	assert.Equal(t, 42, exitCode, fmt.Sprintf("Expected exit code of 42 for container; got %d", exitCode))
 }
+
+// TestServerEndpointValidator tests the workflow of task server endpoint
+func TestServerEndpointValidator(t *testing.T) {
```
Comment (samuelkarp) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Is there a particular reason you hand-wrote this code instead of using the simpletests package?  I would recommend using that instead.

File: misc/task-server-endpoint-validator/Dockerfile (hunk 1 above)
Comment (samuelkarp) [inline misc/task-server-endpoint-validator/Dockerfile]: If you use an image like `busybox` which already includes `wget` or an image like `amazonlinux:2` which already includes `curl`, you can avoid having to construct a new image and instead can just override the command.

File: agent/handlers/task_server_setup.go (hunk 4 above)
Comment (fierlion) [reply agent/handlers/task_server_setup.go]: I see ```w.WriteHeader(http.StatusMovedPermanently)``` aka 301.  Cool.

File: agent/functional_tests/testdata/taskdefinitions/task-server-endpoint-validator/task-definition.json (hunk 3 above)
Comment (shubham2892) [reply agent/functional_tests/testdata/taskdefinitions/task-server-endpoint-validator/task-definition.json]: Ah what about something like task_server_endpoint_validator_container? 🤔

File: misc/task-server-endpoint-validator/Dockerfile (hunk 1 above)
Comment (suneyz) [reply misc/task-server-endpoint-validator/Dockerfile]: awesome! will change it to al2 image

File: scripts/setup-test-registry [hunk 6]
```diff
@@ -45,7 +45,7 @@ mirror_local_image() {
   docker rmi $2
 }
 
-for image in "amazon/amazon-ecs-netkitten" "amazon/amazon-ecs-volumes-test" "amazon/amazon-ecs-pid-namespace-test" "amazon/amazon-ecs-ipc-namespace-test" "amazon/squid" "amazon/awscli" "amazon/image-cleanup-test-image1" "amazon/image-cleanup-test-image2" "amazon/image-cleanup-test-image3" "amazon/fluentd" "amazon/amazon-ecs-agent-introspection-validator" "amazon/amazon-ecs-taskmetadata-validator" "amazon/amazon-ecs-v3-task-endpoint-validator" "amazon/amazon-ecs-container-metadata-file-validator" "amazon/amazon-ecs-elastic-inference-validator"; do
+for image in "amazon/amazon-ecs-netkitten" "amazon/amazon-ecs-volumes-test" "amazon/amazon-ecs-pid-namespace-test" "amazon/amazon-ecs-ipc-namespace-test" "amazon/squid" "amazon/awscli" "amazon/image-cleanup-test-image1" "amazon/image-cleanup-test-image2" "amazon/image-cleanup-test-image3" "amazon/fluentd" "amazon/amazon-ecs-agent-introspection-validator" "amazon/amazon-ecs-taskmetadata-validator" "amazon/amazon-ecs-v3-task-endpoint-validator" "amazon/amazon-ecs-container-metadata-file-validator" "amazon/amazon-ecs-elastic-inference-validator" "amazon/task-server-endpoint-validator"; do
```
Comment (tommyhahn) [inline scripts/setup-test-registry]: Maybe it is trivial but I am wondering why this diff is the only line that specific changes are not highlighted in the deeper green for "amazon/task-server-endpoint-validator"? Or just it happens once in a while...

Comment (suneyz) [reply scripts/setup-test-registry]: not sure, maybe it is a github bug lol

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 5 above)
Comment (sharanyad) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: I believe it is for setting the task iam roles configuration.

**Comment by sharanyad** [comment]
could you please shorten the commit message to 75 characters or less? 
You could a description in a different line below if needed. 
Example commit message: https://github.com/aws/amazon-ecs-agent/pull/993/commits/c140bd39db9fa89c39dbefbaae6cc090d4b152ab

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 5 above)
Comment (suneyz) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: Discussed with Sam offline, we might not need network mode setup for it to work. 

But I'm thinking for now, let's keep test file as hand-written to release it today to solve our customers' issue. I will keep troubleshooting to see if I can make functional test work with simpletests file

**Comment by suneyz** [comment]
> could you please shorten the commit message to 75 characters or less?
> You could a description in a different line below if needed.
> Example commit message: [c140bd3](https://github.com/aws/amazon-ecs-agent/commit/c140bd39db9fa89c39dbefbaae6cc090d4b152ab)

Will do

**Comment by suneyz** [comment]
failed on flaky test TestDockerStopTimeout for both linux and arm. merge the change

**Comment by aaithal** [comment]
> ailed on flaky test TestDockerStopTimeout for both linux and arm. merge the change

Is there an issue tracking that? How long can we expect these tests to be flaky?

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 5 above)
Comment (samuelkarp) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: Task IAM roles are supported with "host", "bridge", and "awsvpc" network modes, and we have existing functional tests that validate this.  There should not be a dependency on using the "host" network mode.

**Comment by suneyz** [comment]
> > failed on flaky test TestDockerStopTimeout for both linux and arm. merge the change
> 
> Is there an issue tracking that? How long can we expect these tests to be flaky?

It is tracked in our sprint board.

**Comment by aaithal** [comment]
> It is tracked in our sprint board.

Can you please create an issue in this repo because that's where we track these things? That way, you can tie that into releases, track when it surfaced, when it was fixed etc. Also, as a customer, or as someone who doesn't have visibility into your sprint board, it's hard to track if there's nothing in the github repo related to that.

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 5 above)
Comment (suneyz) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: Read some document, agree on this should not depend on host mode. I will dive deep to see if my host needs some config to make it functional test on other modes, and I will follow up with another pr to address it once i get it figured out.

**Comment by suneyz** [comment]
> > It is tracked in our sprint board.
> 
> Can you please create an issue in this repo because that's where we track these things? That way, you can tie that into releases, track when it surfaced, when it was fixed etc. Also, as a customer, or as someone who doesn't have visibility into your sprint board, it's hard to track if there's nothing in the github repo related to that.

Github issue created. https://github.com/aws/amazon-ecs-agent/issues/1855
